quam/components/ports.py

```diff
@@ -0,0 +1,266 @@
+                    f"have entry {self.port_type}s for port {self.port}"
+                )
+
+        controller_cfg = config["controllers"].setdefault(controller_name, {})
+        fems_cfg = controller_cfg.setdefault("fems", {})
+        fem_cfg = fems_cfg.setdefault(fem, {})
+        ports_cfg = fem_cfg.setdefault(f"{self.port_type}s", {})
+        port_cfg = ports_cfg.setdefault(port, {})
+        return port_cfg
+
+
+# --- Analog ports --- #
+@quam_dataclass
+class LFAnalogOutputPort(QuamComponent, ABC):
+    port_type: ClassVar[str] = "analog_output"
+
+    offset: float = 0.0
+    delay: int = 0
+    crosstalk: Dict[int, float] = field(default_factory=dict)
+    feedforward_filter: List[float] = field(default_factory=list)
+    feedback_filter: List[float] = field(default_factory=list)
+    shareable: bool = False
+
+    def get_port_properties(self):
+        port_properties = {
+            "delay": self.delay,
+            "crosstalk": self.crosstalk,
+            "feedforward_filter": self.feedforward_filter,
+            "feedback_filter": self.feedback_filter,
+            "shareable": self.shareable,
+        }
+        if self.offset is not None:
+            port_properties["offset"] = self.offset
+        return port_properties
+
+
+@quam_dataclass
+class LFAnalogInputPort(QuamComponent, ABC):
+    port_type: ClassVar[str] = "analog_input"
+
+    offset: float = 0.0
+    gain_db: int = 0
+    shareable: bool = False
+
+    def get_port_properties(self):
+        return {
+            "offset": self.offset,
+            "gain_db": self.gain_db,
+            "shareable": self.shareable,
+        }
+
+
+@quam_dataclass
+class OPXPlusAnalogOutputPort(LFAnalogOutputPort, OPXPlusPort):
+    pass
+
+
+@quam_dataclass
+class OPXPlusAnalogInputPort(LFAnalogInputPort, OPXPlusPort):
+    pass
+
+
+@quam_dataclass
+class LFFEMAnalogOutputPort(LFAnalogOutputPort, FEMPort):
+    sampling_rate: float = 1e9  # Either 1e9 or 2e9
+    upsampling_mode: Literal["mw", "pulse"] = "mw"
+    output_mode: Literal["direct", "amplified"] = "direct"
+
+    def get_port_properties(self) -> Dict[str, Any]:
+        port_properties = super().get_port_properties()
+        port_properties["sampling_rate"] = self.sampling_rate
+        port_properties["upsampling_mode"] = self.upsampling_mode
+        port_properties["output_mode"] = self.output_mode
+        return port_properties
+
+
+@quam_dataclass
+class LFFEMAnalogInputPort(LFAnalogInputPort, FEMPort):
+    sampling_rate: float = 1e9  # Either 1e9 or 2e9
+
+    def get_port_properties(self) -> Dict[str, Any]:
+        port_properties = super().get_port_properties()
+        port_properties["sampling_rate"] = self.sampling_rate
+        return port_properties
+
+
+@quam_dataclass
+class MWFEMAnalogOutputPort(FEMPort):
+    port_type: ClassVar[str] = "analog_output"
+
+    band: int
+    upconverter_frequency: Optional[float] = None
+    upconverters: Optional[Dict[int, float]] = None
```

Comment:
These two are optional, but it's a `one_of`, meaning exactly one must be supplied (not none of them and not both.)

Reply:
Agreed, but I don't think there's a way to specify this in the type. I could add a check in `apply_to_config` and raise an error if none are specified

Reply:
It doesn't have to be in the type.
We need to add a check for that and raise an error (if none exist, or if both exist).
We can also relay on QUA config errors (but that will be less informative to the user).

I would argue that the check need to be right at the beginning of the init, but leaving this up to you.

Reply:
Added a check and error if both are not added